quic/s2n-quic-core/src/crypto/tls/offload.rs

```diff
@@ -0,0 +1,632 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    application,
+    crypto::{
+        tls::{self, NamedGroup, TlsSession},
+        CryptoSuite,
+    },
+    sync::spsc::{channel, Receiver, SendSlice, Sender},
+    transport,
+};
+use alloc::{boxed::Box, collections::vec_deque::VecDeque, sync::Arc, vec::Vec};
+use core::{any::Any, future::Future, task::Poll};
+use std::sync::Mutex;
+
+/// Trait used for spawning async tasks corresponding to TLS operations. Each task will signify TLS work
+/// that needs to be done per QUIC connection.
+pub trait Executor {
+    fn spawn(&self, task: impl Future<Output = ()> + Send + 'static);
+}
+
+/// Allows access to the TlsSession on handshake failure and when the exporter secret is ready.
+pub trait ExporterHandler {
+    fn on_tls_handshake_failed(&self, session: &impl TlsSession) -> Option<Box<dyn Any + Send>>;
+    fn on_tls_exporter_ready(&self, session: &impl TlsSession) -> Option<Box<dyn Any + Send>>;
+}
```

Comment:
It might be kinda nice to add `impl ExporterHandler for ()` in case the application doesn't need this functionality.